crates/vm/src/arch/vm.rs

```diff
@@ -110,70 +122,92 @@ where
         exe: impl Into<VmExe<F>>,
         input: impl Into<Streams<F>>,
     ) -> Result<Vec<ExecutionSegment<F, VC>>, ExecutionError> {
+        let mem_config = self.config.system().memory_config;
         let exe = exe.into();
-        let streams = input.into();
+        let mut streams = input.into();
         let mut segments = vec![];
-        let mem_config = self.config.system().memory_config;
+        let mut memory = AddressMap::from_iter(
+            mem_config.as_offset,
+            1 << mem_config.as_height,
+            1 << mem_config.pointer_max_bits,
+            exe.init_memory.clone(),
+        );
+        let mut pc = exe.pc_start;
+        let mut segment_idx = 0;
+
+        loop {
+            let mut one_segment_result =
+                self.execute_until_segment(exe.clone(), memory, streams, segment_idx, pc)?;
+            streams = one_segment_result.segment.chip_complex.take_streams();
```

Comment:
this is unrelated, but could you see if you can take the cycle tracker from the previous segment and pass it to the next segment? currently it's causing multi-segment flamegraphs to break